src/Type/IntegerRangeType.php

```diff
@@ -220,6 +232,136 @@ public function toBoolean(): BooleanType
 		return new ConstantBooleanType(false);
 	}
 
+	/**
+	 * Return the union with another type, but only if it can be expressed in a simpler way than using UnionType
+	 *
+	 * @param Type $otherType
+	 * @return Type|null
+	 */
+	public function simplifiedUnion(Type $otherType): ?Type
+	{
+		if ($otherType instanceof self || $otherType instanceof ConstantIntegerType) {
+			if ($otherType instanceof self) {
+				$otherMin = $otherType->min;
+				$otherMax = $otherType->max;
+			} else {
+				$otherMin = $otherType->getValue();
+				$otherMax = $otherType->getValue();
+			}
+
+			if (self::isDisjoint($this->min, $this->max, $otherMin, $otherMax, false)) {
+				return null;
+			}
+
+			return self::fromInterval(
+				$this->min !== null && $otherMin !== null ? min($this->min, $otherMin) : null,
+				$this->max !== null && $otherMax !== null ? max($this->max, $otherMax) : null
+			);
+		}
+
+		if (get_class($otherType) === parent::class) {
+			return $otherType;
+		}
+
+		return null;
+	}
+
+	/**
+	 * Return the intersection with another type, but only if it can be expressed in a simpler way than using
+	 * IntersectionType
+	 *
+	 * @param Type $otherType
+	 * @return Type|null
+	 */
+	public function simplifiedIntersection(Type $otherType): ?Type
+	{
+		if ($otherType instanceof self || $otherType instanceof ConstantIntegerType) {
+			if ($otherType instanceof self) {
+				$otherMin = $otherType->min;
+				$otherMax = $otherType->max;
+			} else {
+				$otherMin = $otherType->getValue();
+				$otherMax = $otherType->getValue();
+			}
+
+			if (self::isDisjoint($this->min, $this->max, $otherMin, $otherMax, false)) {
+				return new NeverType();
+			}
+
+			if ($this->min === null) {
+				$newMin = $otherMin;
+			} elseif ($otherMin === null) {
+				$newMin = $this->min;
+			} else {
+				$newMin = max($this->min, $otherMin);
+			}
+
+			if ($this->max === null) {
+				$newMax = $otherMax;
+			} elseif ($otherMax === null) {
+				$newMax = $this->max;
+			} else {
+				$newMax = min($this->max, $otherMax);
+			}
+
+			return self::fromInterval($newMin, $newMax);
+		}
+
+		if (get_class($otherType) === parent::class) {
+			return $this;
+		}
+
+		return null;
+	}
+
+	/**
+	 * Return the different with another type, or null if it cannot be represented.
+	 *
+	 * @param Type $typeToRemove
+	 * @return Type|null
+	 */
+	public function simplifiedDifference(Type $typeToRemove): ?Type
```

Comment:
What about `removeSelf`?

Reply:
I agree that the names aren't optimal. But `removeSelf` sounds like it would work the other way around. It's not the self being removed. It's the other type being removed from self. I chose "simplified" to stress the fact that it only performs the operation when it can be represented in a simple way. Other code is still supposed to use `TypeCombinator`. And `TypeCombinator` will then in turn call these methods on the types, trying whatever works, e.g. `$a->union(b) ?? $b->union($a) ?? UnionType($a, $b)`.

About the names, I think using "self" is redundant, since when calling a method, the self is always involved anyway. Perhaps `tryUnion()`, `tryIntersect()`, `tryRemove()`?

btw, I only intended the 3 type combination methods to be moved to `Type`, not `shift()`. Although that's possible too of course (with a different name). Or perhaps it should only be moved up to `IntegerType`, because it really doesn't make sense for other types.

Reply:
Yeah, the `try` variants make sense to me :) If you rename those methods, I'm gonna merge this.

> Or perhaps it should only be moved up to IntegerType

This wouldn't help me with my long term goal of making all `instanceof *Type` illegal. And it also wouldn't cover the case where you have more `IntegerRangeType|IntegerRangeType` in a UnionType :)